### PR TITLE
Import man pages

### PR DESCRIPTION
**Pull request by tiborvass**
Imports the man/ folder from moby/moby, preserving history.

Reuses main Dockerfile to generate man pages

Removes the use of glide, and uses vndr instead.

947efe9 is the merge commit that imports the history
019a052 is the commit that makes the relevant changes to make it work.

Run `make -f docker.Makefile manpages` to test

### CODE REVIEW

**Comment by andrewhsu** [comment]
LGTM

stuff works:
```
$ make -f docker.Makefile manpages
$ ls man/man*
man/man1:
docker-attach.1             docker-exec.1               docker-node-rm.1            docker-service-update.1
docker-build.1              docker-export.1             docker-node-update.1        docker-service.1
docker-checkpoint-create.1  docker-history.1            docker-node.1               docker-stack-deploy.1
docker-checkpoint-ls.1      docker-image-build.1        docker-pause.1              docker-stack-ls.1
docker-checkpoint-rm.1      docker-image-history.1      docker-plugin-create.1      docker-stack-ps.1
docker-checkpoint.1         docker-image-import.1       docker-plugin-disable.1     docker-stack-rm.1
docker-commit.1             docker-image-inspect.1      docker-plugin-enable.1      docker-stack-services.1
docker-container-attach.1   docker-image-load.1         docker-plugin-inspect.1     docker-stack.1
docker-container-commit.1   docker-image-ls.1           docker-plugin-install.1     docker-start.1
docker-container-cp.1       docker-image-prune.1        docker-plugin-ls.1          docker-stats.1
docker-container-create.1   docker-image-pull.1         docker-plugin-push.1        docker-stop.1
docker-container-diff.1     docker-image-push.1         docker-plugin-rm.1          docker-swarm-init.1
docker-container-exec.1     docker-image-rm.1           docker-plugin-set.1         docker-swarm-join-token.1
docker-container-export.1   docker-image-save.1         docker-plugin-upgrade.1     docker-swarm-join.1
docker-container-inspect.1  docker-image-tag.1          docker-plugin.1             docker-swarm-leave.1
docker-container-kill.1     docker-image.1              docker-port.1               docker-swarm-unlock-key.1
docker-container-logs.1     docker-images.1             docker-ps.1                 docker-swarm-unlock.1
docker-container-ls.1       docker-import.1             docker-pull.1               docker-swarm-update.1
docker-container-pause.1    docker-info.1               docker-push.1               docker-swarm.1
docker-container-port.1     docker-inspect.1            docker-rename.1             docker-system-df.1
docker-container-prune.1    docker-kill.1               docker-restart.1            docker-system-events.1
docker-container-rename.1   docker-load.1               docker-rm.1                 docker-system-info.1
docker-container-restart.1  docker-login.1              docker-rmi.1                docker-system-prune.1
docker-container-rm.1       docker-logout.1             docker-run.1                docker-system.1
docker-container-run.1      docker-logs.1               docker-save.1               docker-tag.1
docker-container-start.1    docker-network-connect.1    docker-search.1             docker-top.1
docker-container-stats.1    docker-network-create.1     docker-secret-create.1      docker-unpause.1
docker-container-stop.1     docker-network-disconnect.1 docker-secret-inspect.1     docker-update.1
docker-container-top.1      docker-network-inspect.1    docker-secret-ls.1          docker-version.1
docker-container-unpause.1  docker-network-ls.1         docker-secret-rm.1          docker-volume-create.1
docker-container-update.1   docker-network-prune.1      docker-secret.1             docker-volume-inspect.1
docker-container-wait.1     docker-network-rm.1         docker-service-create.1     docker-volume-ls.1
docker-container.1          docker-network.1            docker-service-inspect.1    docker-volume-prune.1
docker-cp.1                 docker-node-demote.1        docker-service-logs.1       docker-volume-rm.1
docker-create.1             docker-node-inspect.1       docker-service-ls.1         docker-volume.1
docker-deploy.1             docker-node-ls.1            docker-service-ps.1         docker-wait.1
docker-diff.1               docker-node-promote.1       docker-service-rm.1         docker.1
docker-events.1             docker-node-ps.1            docker-service-scale.1

man/man5:
Dockerfile.5         docker-config-json.5

man/man8:
dockerd.8
```

File: Makefile [hunk 1]
```diff
@@ -11,12 +11,13 @@ build: clean
...
 clean:
 	@rm -rf ./build/*
+	@rm -rf ./man/man[1-9]
 
 # run go test
 # the "-tags daemon" part is temporary
 .PHONY: test
 test:
-	@go test -tags daemon -v $(shell go list ./... | grep -v /vendor/)
+	@go test -tags daemon -v $(shell go list ./... | grep -vE '/vendor/|github.com/docker/cli/man$$')
```
Comment (dnephin) [inline Makefile]: Why is it necessary to exclude the man directory from testing?

File: gometalinter.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
   "Vendor": true,
   "Deadline": "2m",
   "Sort": ["linter", "severity", "path"],
-  "Exclude": ["cli/compose/schema/bindata.go"],
+  "Exclude": ["cli/compose/schema/bindata.go", "man/generate.go"],
```
Comment (dnephin) [inline gometalinter.json]: Why is this necessary? I don't think we should exclude it from linting

File: Makefile [hunk 3]
```diff
@@ -32,10 +33,15 @@ cross: clean
 .PHONY: vendor
 vendor: vendor.conf
 	@vndr 2> /dev/null
-	@script/validate/check-git-diff vendor
+	@scripts/validate/check-git-diff vendor
+
+## Generate man pages from go source and markdown
+.PHONY: manpages
+manpages:
+	@man/generate.sh
```
Comment (dnephin) [inline Makefile]: This doesn't need to be `.PHONY`, this is one of the rare cases when a Makefile is actually appropriate.

I believe something like this would be correct:

```
man/man*: man/
    @man/generate.sh

manpages: man/man*
```

File: man/generate.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	go install "$MD2MAN_REPO"
+)
+
+VENDOR_MD5="$(md5sum vendor.conf)"
+cp vendor.conf /tmp/vendor.conf
+cp man/vendor.tmp vendor.conf
+
+grep -v '^#' man/vendor.tmp | while read dep; do
+	vndr $(echo "$dep" | cut -d' ' -f1)
+done
```
Comment (dnephin) [inline man/generate.sh]: There has to be a better way to do this.

Why not keep this as a separate Dockerfile? Now that we have ARG in FROM I think we can have a single file.

File: gometalinter.json (hunk 2 above)
Comment (tiborvass) [inline gometalinter.json]: Because packages are missing when linting.

File: Makefile (hunk 1 above)
Comment (tiborvass) [inline Makefile]: Because of missing packages, that are included only after the vendor-on-the-fly part of generate.sh

File: Makefile (hunk 3 above)
Comment (ijc25) [inline Makefile]: Can you use wildcards in `Makefile` rules like that? I didn't think so, you would need [`$(wildcard man/man*)`](https://www.gnu.org/software/make/manual/html_node/Wildcard-Function.html), I think.

**But** be careful: this:
```
foo bar: baz
     make-foo-and-bar baz
```
Does not cause `make-foo-and-bar baz` to run once whenever `baz` changes. It will be run once for `foo` and once for `bar`. That is unlikely to be what you wanted, at least judging from the lack of parameters to `man/generate.sh` in your example (if it had been passed `$@` then it might be correct).

Comment (dnephin) [inline Makefile]: I guess this would be easier to do if we wrote the man pages to `man/dist/` then the target would be just that string without the need for wildcards.

Comment (ijc25) [inline Makefile]: Not entirely sure how `make` deals with targets or dependencies which are directories rather then files, but if that sort of thing works (I guess it just `stat`s and uses the `mtime` just like for a file), sure.

Comment (dnephin) [inline Makefile]: Yes, I think that's how it works. I did it for `vendor` (in this Makefile) and it seems to work so far.

**Comment by vdemeester** [comment]
@tiborvass needs a rebase 👼
